simln-lib: add htlc interceptor for simulated nodes

elnosh opened the pull request:
for https://github.com/bitcoin-dev-project/sim-ln/issues/255

Mostly took changes from these 2 commits to add an `Interceptor` trait:
- https://github.com/carlaKC/sim-ln/commit/c5d7649d0969f524dc1076760128338f38853ed8 
- https://github.com/carlaKC/sim-ln/commit/228706ca3ef70d6111535ae75ef09af724dea448


 Some of the things I changed from those commits:
- `intercept_htlc` now returns a `Result` to communicate the result of the intercepted htlc instead of sending it through a channel.
- spawns a task in a `JoinSet` for each interceptor's `intercept_htlc` so that if any of those holds the htlc for a long time it does not block the other interceptors. It then waits for the completion of the tasks in the joinset. If any of them returns a result to fail the htlc, it drops the `JoinSet` since there is no need to wait for completion of other tasks because the htlc will fail anyways.

carlaKC commented:
Driveby comment: could we split up the logic adding a unique index and the interceptor into separate commits?
For the sake of breaking up into smaller logical chunks / reviewability.

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -930,16 +1074,99 @@ async fn add_htlcs(
+                );
+
+                let interceptor_clone = Arc::clone(interceptor);
+                intercepts.spawn(async move { interceptor_clone.intercept_htlc(request).await });
```
This might be a good spot for `futures::future::join_all`?

Rather than needing to spawn a new task here, we can just call (but not await) the `intercept_htlc` method and then await them in whatever order they arrive after that?

Author's choice. - I don't see us running with _so many_ interceptors that creating these tasks starts to give us a performanc hit.

elnosh replied:
looked into `join_all` but if iiuc it will wait until all futures resolve. I think we'd prefer to get the responses as the tasks finish rather than waiting for all to finish? that way we can fail the other interceptors if one failed the htlc.

carlaKC replied:
> I think we'd prefer to get the responses as the tasks finish rather than waiting for all to finish? that way we can fail the other interceptors if one failed the htlc.

Yeah indeed, we want to hear back from the first completing one so that we can cancel the rest. Could use `FuturesUnordered` but I think that's an over-optimization (it's for lots of futures), so happy to leave as-is 👍

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -930,16 +1074,99 @@ async fn add_htlcs(
+                        }
+                    },
+                    Ok(Err(e)) => {
+                        drop(intercepts);
```
I think that we may need a more gentle shutdown than dropping the task (which afaik will force abort the `intercept_htlc` call).

I can see it being difficult to write interceptor code that works with another system when your code may be run half way and then killed - could end up with funny states. My instinct is to provides a `triggered` pair and instruct interceptors to listen on it for shutdown signals? Open to other approaches as well.

I do think that the ability to shut down all the other interceptors once the first error is reached is a really, really nice feature.

elnosh replied:
sounds good :thumbsup: provide that trigger in the `InterceptRequest` and send on it if we need to fail other interceptors? 
just to note, although `intercept_htlc` may force-aborted, interceptor should still get notified about htlc resolution through `notify_resolution` but I think we'd still like to shutdown gently through `intercept_htlc`?

carlaKC replied:
> provide that trigger in the InterceptRequest and send on it if we need to fail other interceptors?

I think that we should handle the triggering? 
- Give each interceptor a listener
- On first interceptor error, trigger shutdown

Then interceptors are responsible for listening on that shutdown listener and getting their state in order before they exit. We still wait for each to finish, but that should be relatively quick because we've signaled that it's shutdown time.

> Interceptor should still get notified about htlc resolution through notify_resolution

As-is I don't think we'd notify if the HTLC is failed by the interceptor? It'll never be "fully" forwarded by the node, so we don't notify it being resolved.

elnosh replied:
> * Give each interceptor a listener 
> * On first interceptor error, trigger shutdown
> 
> 
> Then interceptors are responsible for listening on that shutdown listener and getting their state in order before they exit. We still wait for each to finish, but that should be relatively quick because we've signaled that it's shutdown time.

got it, will do. 

> As-is I don't think we'd notify if the HTLC is failed by the interceptor? It'll never be "fully" forwarded by the node, so we don't notify it being resolved.

I thought we would. Because if an interceptor returns an error, we then return it in `add_htlcs` and then call `remove_htlcs` here: https://github.com/elnosh/sim-ln/blob/2de405fdcb37c8ffe4bd8d2cc0077ef7099cc3ed/simln-lib/src/sim_node.rs#L1274 which internally calls `notify_resolution` for each interceptors

carlaKC replied:
> I thought we would.

Ok cool, forgot when that is/isn't called bu sgtm!

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -655,6 +668,126 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
+pub trait Interceptor: Send + Sync {
+    /// Implemented by HTLC interceptors that provide input on the resolution of HTLCs forwarded in the simulation.
+    async fn intercept_htlc(&self, req: InterceptRequest)
+        -> Result<CustomRecords, ForwardingError>;
```
If we only have one layer of `Result` here, how does the `intercept_htlc` call tell the simulator that it's hit a ciritical error and wants to shut down (rather than a forwarding error for this payment specifically)?

elnosh replied:
I wasn't sure if that's something we'd want. I didn't like the idea of nested `Result`s. It could instead be communicated through specific variants in `ForwardingError`? Simulation could check with `is_critical` if the `ForwardingError` returned warrants a shutdown.

carlaKC replied:
> It could instead be communicated through specific variants in ForwardingError

That seems like a bit of a layering violation to me. An`InterceptorError` for interceptors that might want to fail forwards with their own `ForwardingFalure` reason that isn't defined in simln (eg, you don't have enough reputation) makes sense to me, but we're stretching its definition a bit to cover things like unexpected internal state errors.

Nested results are definitely ugly, but I think that it more clearly represents the two types of failures that we could have here? One for "fail this HTLC", one for "something has gone wrong with my interception". If we define a type for the inner result the function signatures won't be too ugly at least.

elnosh replied:
fair! will change 👍

carlaKC reviewed (commented):
Main comments are around how we handle shutdown in a reaonsable way. Specifically:
* An interceptor has had a critical failure, how does it tell the simulator to shut down
* The simultor needs to terminate the interceptors, how can we cleanly do this so that interception code isn't left in a bad state

elnosh commented:
> * An interceptor has had a critical failure, how does it tell the simulator to shut down
> 

I wasn't sure if letting the interceptor shutdown the simulation was something we'd want but it could be done through specific variants in the `ForwardingError` returned? Simulation could check with `is_critical` if the ForwardingError returned warrants a shutdown. Although `Interceptor`s should be aware of this to trigger a shutdown by providing specific variants. 

> * The simultor needs to terminate the interceptors, how can we cleanly do this so that interception code isn't left in a bad state

should we include a channel or shutdown signal in `InterceptRequest`?

carlaKC commented:
WDYT about adapting [this latency interceptor](https://github.com/carlaKC/sim-ln/commit/2a35bd55b5da2e2cc73b1ed8b03f84bf39613609) and [surfacing](https://github.com/carlaKC/sim-ln/commit/71bd12e29c029f19e987ae91c818e645e373b309) it on the CI? 

Nice to have a user of the interceptor that we can run + test in master rather than merging in some unused code and having to test it externally.

elnosh commented:
sgtm, as I was doing exactly that and testing externally heh.

elnosh commented on simln-lib/src/sim_node.rs:
```diff
@@ -643,6 +668,126 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
+
+    /// Notifies the interceptor that a previously intercepted htlc has been resolved. Default implementation is a no-op
+    /// for cases where the interceptor only cares about interception, not resolution of htlcs.
+    async fn notify_resolution(
+        &self,
+        _res: InterceptResolution,
+    ) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+        Ok(())
+    }
```
rethinking this, I'm wondering what type of error should this be or if this method should return a `Result` at all? Leaning towards this should not return a `Result` since the simulation is just notifying the interceptor of a resolution.

carlaKC replied:
Perhaps the interceptor wants to signal shutdown on the `notify_resolution` because it got information it wasn't expecting? Or a database became unavailable, etc.

elnosh commented:
Added changes addressing comments. Using nested results now to differentiate between normal and critical errors that could be returned. If a `CriticalError` is received, it returns and triggers a shutdown. If a `ForwardingError` happens during interception, we send a signal to other interceptors to let them know they should shut down. 

Also took your commit https://github.com/carlaKC/sim-ln/commit/2a35bd55b5da2e2cc73b1ed8b03f84bf39613609 as suggested to add the latency interceptor. 

If the simulation runs with the `LatencyInterceptor` it can log some error messages during shutdown 
```
2025-05-14T15:59:40.049Z ERROR [simln_lib] Track payment failed for 7a56bab4c9b27d3cc09438dd6c32255eb4ecb4559337e9125c8920f8b071ab61: Track payment error: shutdown during payment tracking.
```
because we triggered a shutdown while an intercepted payment (with the added latency) hasn't resolved.

carlaKC reviewed (commented):
This is looking really good! Only two major comments on shutdown + cli option for latency.

Been a bit nitty about docs because this feature really will only be used when people use simln as a library so it's important to get the docs readable.

elnosh commented on simln-lib/src/sim_node.rs:
```diff
@@ -980,15 +1102,104 @@ async fn add_htlcs(
+            // the HTLC. If any of the interceptors did return an error, we send a shutdown signal
+            // to the other interceptors that may have not returned yet.
+            let mut interceptor_failure = None;
+            while let Some(res) = intercepts.join_next().await {
```
> I think that we can do this by selecting on `intercepts.join_next()` and pass in `SimGraph`'s `shutdown_trigger` and shut down the interceptors if we get the high level signal that it's time to shut down.

I think we'll need to pass in a shutdown listener from the upstream `Simulation` since currently we don't have a way to listen for a shutdown signal in `SimGraph` afaik. `shutdown_trigger` is the one we use to trigger on critical errors.


> We're somewhat running into a flaw with `triggered` here - that we can't create a child trigger that would shut down with the parent, but maybe that's a feature not a bug. Either way, I think it's outside of the scope of this PR to rework all that so I think a `select` is our best option (even if ugly).

maybe for another PR but could look into [CancellationToken](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.child_token) looks like it lets you create "child tokens" that can get cancelled along with the parent

carlaKC replied:
> I think we'll need to pass in a shutdown listener from the upstream Simulation since currently we don't have a way to listen for a shutdown signal in SimGraph afaik.

Right, we need `shutdown_listener` 🤦‍♀️   SGTM, we can add it to `SimGraph` and then pass it along.

> maybe for another PR but could look into [CancellationToken](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.child_token) looks like it lets you create "child tokens" that can get cancelled along with the parent

Yeah I like the idea of investigating this - took a look when we updated the leaky joinset and liked it.

elnosh replied:
moved the htlc interception logic to a separate method. I didn't end up needing to use `select` here since we are passing the shutdown listener to `SimGraph` now so it can be passed directly in the `InterceptRequest` to the interceptors.

elnosh commented on simln-lib/src/sim_node.rs:
```diff
@@ -980,15 +1102,104 @@ async fn add_htlcs(
+                                    e.insert(v);
+                                },
+                            };
+                        }
```
Since `CustomRecords` is just a alias for `HashMap`, I ended up leaving as is instead of changing because I would need to wrap the `HashMap` in a struct to define a method on it so I thought it would be ok to just leave it. Lmk if you'd prefer to create type for and I can change it.

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -692,6 +704,178 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
+                        match attached_custom_records.entry(k) {
+                            Entry::Occupied(e) => {
+                                let existing_value = e.get();
+                                if *existing_value == v {
+                                    return Err(CriticalError::DuplicateCustomRecord(k));
+                                }
```
Shouldn't we fail if `existing_value != v`? The same value means we can merge with no conflicts, different values means we've got an invalid result that we don't know how to handle.

elnosh replied:
ah, I misunderstood the "conflict" case we wanted to fail. Thought that if they were exactly the same we'd want to fail. But that doesn't make much sense :man_facepalming: will change

carlaKC replied:
TBH I've never found an intuitive way to do this type of merge API - a world where each interceptor is only allowed to set one value also _sort of_ makes sense, but it's just more restrictive.

But yeah, let's go the more permissive "error on conflict" approach 👍

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -692,6 +704,178 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
+        // to the other interceptors that may have not returned yet.
+        let mut interceptor_failure = None;
+        while let Some(res) = intercepts.join_next().await {
+            let res =
+                res.map_err(|join_err| CriticalError::InterceptorError(join_err.to_string()))?;
```
Wondering about how best to handle `CriticalError` but here and below for `DuplicateCustomRecords`.

When we error out, will `shutdown_trigger` be called somewhere up the stack? Otherwise the other interceptors don't get the signal to shut down. 

I think that the strictly cleanest way to do this would be to:
1. Track that we ran into a critical error
2. Trigger shutdown
3. Loop through remaining tasks to make sure they completed
4. Return the first critical error we hit

Doesn't need to be in this PR, but maybe let's write this up in an issue just so that we don't forget?

carlaKC replied:
Yeah this is very nitpickey - in the case of critical shutdown a few interceptors might get aborted because the joinset is dropped (this will actually happen before shutdown is triggered, so my previous comment about it getting triggered up the stack doesn't really matter).

Let's leave as-is and make an issue noting that we don't guarantee clean shutdown of interceptors when we hit a `CriticalError`. I don't see it being much of a problem in practice, but good to have an issue so it's known at least.

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -980,16 +1189,48 @@ async fn add_htlcs(
+            outgoing_amount_msat: outgoing_amount - hop.fee_msat,
+            incoming_expiry_height: outgoing_cltv,
+            outgoing_expiry_height: outgoing_cltv - hop.cltv_expiry_delta,
+            shutdown_listener: shutdown_signal_clone.1.clone(),
```
If we use the top level `shutdown_signal` for intercepts and trigger on `FowardingErrors` then won't the simulation shut down any time an interceptor fails a payment? I think that it needs:
- A shutdown trigger that is used by only the interceptors
- A shutdown listener that is used by only the interceptors
- A shutdown listener that is used by the simulation

It's pretty annoying for the interceptor to have two shutdown signals, so we could look at an approach where we monitor the simulation listener and trigger the interceptor one if it's triggered (this is where child triggers would have been great).

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -1053,9 +1313,20 @@ async fn propagate_payment(
+    shutdown_signal: (Trigger, Listener),
 ) {
-    let notify_result = match add_htlcs(nodes.clone(), source, route.clone(), payment_hash).await {
+    let shutdown_clone = (shutdown_signal.0.clone(), shutdown_signal.1.clone());
```
nit: simplify to `let shutdown_clone = shutdown_signal.clone();`?

carlaKC commented on simln-lib/src/sim_node.rs:
```diff
@@ -1896,4 +2198,168 @@ mod tests {
+            result.unwrap().payment_outcome,
+            PaymentOutcome::Unknown
+        ));
+
```
Right now I can make this test fail by adding a `listener` to the test kit and checking that we've triggered top level shutdown of the simulation (because a failed interceptor shuts everything down).

```
		select!{
			_ = test_kit.listener => {assert!(false, "simulation shut down by interceptor")},
			_ = async {} => {},
		}
```

Would be good to add this in above here to assert that interceptor failures aren't killing the whole simulation.

carlaKC reviewed (commented):
Sorry for the slow turnaround on this one!

A few thoughts on tricksey shutdown stuff.

elnosh commented:
Thanks for the patience! Hopefully closer now :fingers_crossed:

From latest feedback, made following changes:
- Created trigger to be used by interceptors only and passing this listener to the `InterceptRequest`.
- `handle_intercepted_htlc` gets a `shutdown_listener` and if we get a signal to shutdown it will trigger the interceptors to shutdown too. 
- If we get a forwarding error we trigger the interceptors to shutdown but since triggers (simulation shutdown and interceptor shutdown) are separated, this should not cause a simulation to shutdown.

elnosh commented on simln-lib/src/sim_node.rs:
```diff
@@ -684,6 +704,205 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
+                    }
+                }
+                _ = shutdown_listener.clone() => {
+                    // If we get a simulation shutdown signal, we trigger the interceptors to
+                    // shutdown and wait for them to finish before returning.
+                    interceptor_trigger.trigger();
+                    while intercepts.join_next().await.is_some() {}
+                    break 'get_resp
+                }
```
one thing I wasn't completely sure was here. If we get simulation shutdown, it triggers the signal to the interceptors and it just waits for them to finish and then break

elnosh commented:
thanks! added debug statement and `is_empty` check
